kbchat/kbchat_test.go

```diff
@@ -1,184 +1,300 @@
+	config := readAndParseTestConfig(t)
+	oneOnOneChannel := Channel{
+		Name: fmt.Sprintf("%s,%s", config.Bots[botName].Username, config.Bots[oneOnOnePartner].Username),
 	}
-	location := strings.TrimSpace(string(out))
-	return location, nil
+	return oneOnOneChannel
 }
 
-func copyFile(source, dest string) error {
-	sourceData, err := ioutil.ReadFile(source)
-	if err != nil {
-		return err
+func getTeamChatChannel(t *testing.T, teamName string) Channel {
+	config := readAndParseTestConfig(t)
+	teamChannel := Channel{
+		Name:        config.Teams[teamName].Name,
+		Public:      false,
+		MembersType: "team",
+		TopicName:   config.Teams[teamName].TopicName,
+		TopicType:   "chat",
 	}
+	return teamChannel
+}
 
-	err = ioutil.WriteFile(dest, sourceData, 0777)
-	if err != nil {
-		return err
-	}
+func testBotTeardown(t *testing.T, bot *API, dir string) {
+	err := bot.Shutdown()
+	require.NoError(t, err)
+	err = os.RemoveAll(dir)
+	require.NoError(t, err)
+}
 
-	return nil
+func getMostRecentMessage(t *testing.T, bot *API, channel Channel) Message {
+	messages, err := bot.GetTextMessages(channel, false)
+	require.NoError(t, err)
+	return messages[0]
 }
 
-// Creates the working directory and copies over the keybase binary in PATH.
-// We do this to avoid any version mismatch issues.
-func prepWorkingDir(workingDir string) (string, error) {
-	kbLocation, err := whichKeybase()
-	if err != nil {
-		return "", err
-	}
+func getConvIDForChannel(t *testing.T, bot *API, channel Channel) string {
+	messages, err := bot.GetTextMessages(channel, false)
+	require.NoError(t, err)
+	convID := messages[0].ConversationID
+	return convID
+}
 
-	err = os.Mkdir(workingDir, 0777)
-	if err != nil {
-		return "", err
-	}
-	kbDestination := path.Join(workingDir, "keybase")
+func TestGetUsername(t *testing.T) {
+	config := readAndParseTestConfig(t)
+	alice, dir := testBotSetup(t, "alice")
+	defer testBotTeardown(t, alice, dir)
+	require.Equal(t, alice.GetUsername(), config.Bots["alice"].Username)
+}
 
-	err = copyFile(kbLocation, kbDestination)
-	if err != nil {
-		return "", err
-	}
+func TestGetConversations(t *testing.T) {
+	alice, dir := testBotSetup(t, "alice")
+	defer testBotTeardown(t, alice, dir)
+	conversations, err := alice.GetConversations(false)
+	require.NoError(t, err)
+	require.True(t, len(conversations) > 0)
+}
 
-	return kbDestination, nil
+func TestGetTextMessages(t *testing.T) {
+	alice, dir := testBotSetup(t, "alice")
+	defer testBotTeardown(t, alice, dir)
+	channel := getOneOnOneChatChannel(t, "alice", "bob")
+	messages, err := alice.GetTextMessages(channel, false)
+	require.NoError(t, err)
+	require.True(t, len(messages) > 0)
 }
 
-func deleteWorkingDir(workingDir string) error {
-	return os.RemoveAll(workingDir)
+func TestSendMessage(t *testing.T) {
+	alice, dir := testBotSetup(t, "alice")
+	defer testBotTeardown(t, alice, dir)
+	channel := getOneOnOneChatChannel(t, "alice", "bob")
```

Comment:
if this were `getOneOnOneChatChannel(t, alice, bob)` then you wouldn't need to parse the config inside that method. and parsing the config is something we should try to do just once at the very beginning. we'll have more flexibility to customize flows if we aren't constantly going back to look at that global state.

Reply:
i think what i'm reacting to here is that, if possible, only testBotSetup should know about the config file. what it is, what it looks like, how to read it, etc.

Reply:
@xgess what args would `getTeamChatChannel` take then?

Reply:
changing this doesn't seem like an obvious improvement. :/